Mathlib/Order/Normal.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Order.SuccPred.InitialSeg
+
+/-!
+# Normal functions
+
+A normal function between well-orders is a strictly monotonic continuous function. Normal functions
+arise chiefly in the context of cardinal and ordinal-valued functions. Unfortunately, Mathlib places
+these earlier in the import chain than the topological notion of continuity.
+
+We instead opt for an alternate but equivalent definition: a normal function is a strictly monotonic
+function `f` such that at successor limits `a`, `f a` is the least upper bound of `f b` with
+`b < a`.
+
+## TODO
+
+* Prove the equivalence with the standard definition (in some other file).
+* Replace `Ordinal.IsNormal` by this more general notion.
+-/
+
+open Order Set
+
+variable {α β γ : Type*} {a b : α} {f : α → β} {g : β → γ}
+
+namespace Order
+
+/-- A normal function between well-orders is a strictly monotonic continuous function. -/
+structure IsNormal [LinearOrder α] [LinearOrder β] (f : α → β) : Prop where
+  strictMono : StrictMono f
+  /-- Combined with strict monotonicity, this condition implies that `f a` is the *least* upper
+  bound for `f '' Iio a`. -/
+  mem_lowerBounds_upperBounds {a : α} (ha : IsSuccLimit a) :
+    f a ∈ lowerBounds (upperBounds (f '' Iio a))
```

Comment:
Is this a more useful definition than using IsLUB? Your docstring here seems to suggest that'd be nicer, at least!

Reply:
This is actually the RHS of the `IsLUB (f a) (f '' Iio a)`. The LHS is `f a ∈ upperBounds (f '' Iio a)`, which is necessarily true by strict monotonicity.

Note that for a normal function, the stronger result `IsNormal.map_isLUB` holds: if `a` is the LUB of `s`, then `f a` is the LUB of `f '' s`.

Reply:
I feel like would be nicer if you stated them in terms of `IsLUB`: It has more API than `lowerBounds (upperBounds _)`!

You can turn the current constructor into a new `IsNormal.of_mem_lowerBounds_upperBounds` constructor, for the few (/no?) cases where `lowerBounds (upperBounds _)` is easier to handle.

Reply:
I really do think this `lowerBounds (upperBounds _)` predicate is easier to work with. Almost always the first step to proving `IsLUB (f a) (f '' Iio a)` is going to be to call `split`, at which point you have two goals, one of which is the `lowerBounds (upperBounds _)` one!

Reply:
> Almost always the first step to proving `IsLUB (f a) (f '' Iio a)` is going to be to call `split`, at which point you have two goals, one of which is the `lowerBounds (upperBounds _)` one!

Really? That's not my experience! Maybe you just didn't find the right API last time you tried proving an `IsLUB` goal? eg `isLUB_sSup`.

Reply:
Hmm. Independently of this, I already proved `IsNormal.of_succ_lt`, which has similar assumptions to what you're asking (it requires a well-ordered domain, but that's the intended use case anyways). Perhaps that's enough?

Reply:
No, I strongly believe that the *default* constructor should be the one spelled using API, unless there's a technical limitation (which is not the case here). Why can't you just swap out the definition?

Reply:
Eh, alright. I've implemented your suggestion.